application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/websocket/WebsocketServer.java

```diff
@@ -0,0 +1,223 @@
+package org.openjdk.jmc.flightrecorder.ui.websocket;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Function;
+import java.util.logging.Level;
+import java.util.stream.Collectors;
+
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.servlet.ServletContextHandler;
+import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.api.WebSocketAdapter;
+import org.eclipse.jetty.websocket.server.config.JettyWebSocketServletContainerInitializer;
+import org.eclipse.jetty.websocket.servlet.WebSocketUpgradeFilter;
+import org.openjdk.jmc.common.item.IItemCollection;
+import org.openjdk.jmc.flightrecorder.serializers.dot.DotSerializer;
+import org.openjdk.jmc.flightrecorder.serializers.json.FlameGraphJsonSerializer;
+import org.openjdk.jmc.flightrecorder.serializers.json.IItemCollectionJsonSerializer;
+import org.openjdk.jmc.flightrecorder.stacktrace.FrameSeparator;
+import org.openjdk.jmc.flightrecorder.stacktrace.FrameSeparator.FrameCategorization;
+import org.openjdk.jmc.flightrecorder.stacktrace.graph.StacktraceGraphModel;
+import org.openjdk.jmc.flightrecorder.stacktrace.tree.StacktraceTreeModel;
+import org.openjdk.jmc.flightrecorder.ui.FlightRecorderUI;
+
+public class WebsocketServer {
+
+	private static int MAX_MESSAGE_SIZE = 1024 * 1024 * 1024;
+	private static int IDLE_TIMEOUT_MINUTES = 5;
+
+	private final int port;
+	private Server server;
+	private List<WebsocketConnectionHandler> handlers = new ArrayList<>();
+	private List<WebsocketConnectionHandler> treeHandlers = new ArrayList<>();
+	private List<WebsocketConnectionHandler> graphHandlers = new ArrayList<>();
```

Comment:
Since there is more than one thread involved into accessing those handlers I would use CopyOnWriteArrayList to be safe on the access.